File: agent/tools/code_exec_contract.py
#
#  Copyright 2026 The InfiniFlow Authors. All Rights Reserved.
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#

from __future__ import annotations

import json
from collections.abc import Mapping
from typing import Any


SYSTEM_OUTPUT_KEYS = frozenset(
    {
        "content",
        "actual_type",
        "_ERROR",
        "_ARTIFACTS",
        "_ATTACHMENT_CONTENT",
        "raw_result",
        "_created_time",
        "_elapsed_time",
    }
)


class ContractError(ValueError):
    pass


def _validate_business_output_name(name: str) -> None:
    if not name or not name.strip():
        raise ContractError("CodeExec business output name must not be empty")
    if name in SYSTEM_OUTPUT_KEYS:
        raise ContractError(f"CodeExec reserved output name is not allowed: {name}")
    if "." in name:
        raise ContractError(f"CodeExec business output name must not contain '.': {name}")


def select_business_output(outputs: Mapping[str, Any]) -> tuple[str, Any]:
    business_outputs = [(name, meta) for name, meta in outputs.items() if name not in SYSTEM_OUTPUT_KEYS]
    if len(business_outputs) != 1:
        raise ContractError(
            f"CodeExec contract must contain exactly one business output, got {len(business_outputs)}"
        )
    _validate_business_output_name(business_outputs[0][0])
    return business_outputs[0]


def normalize_output_value(value: Any) -> Any:
    if isinstance(value, (tuple, list)):
        return [normalize_output_value(item) for item in value]
    if isinstance(value, dict):
        return {key: normalize_output_value(item) for key, item in value.items()}
    return value


def infer_actual_type(value: Any) -> str:
    value = normalize_output_value(value)
    if value is None:
        return "Null"
    if isinstance(value, bool):
        return "Boolean"
    if _is_number(value):
        return "Number"
    if isinstance(value, str):
        return "String"
    if isinstance(value, dict):
        return "Object"
    if isinstance(value, list):
        if not value:
            return "Array<Any>"
        inferred = {infer_actual_type(item) for item in value}
        if len(inferred) == 1:
            return f"Array<{inferred.pop()}>"
        return "Array<Any>"
    return "Any"


def render_canonical_content(value: Any) -> str:
    value = normalize_output_value(value)
    if value is None:
        return ""
    if isinstance(value, str):
        return value
    if isinstance(value, (dict, list)):
        return json.dumps(value, ensure_ascii=False, indent=2, sort_keys=True)
    return str(value)


def _is_number(value: Any) -> bool:
    return isinstance(value, (int, float)) and not isinstance(value, bool)


def _validate_top_level_value_domain(value: Any) -> None:
    allowed = value is None or isinstance(value, (bool, str, dict, list)) or _is_number(value)
    if not allowed:
        raise ContractError(
            f"CodeExec unsupported top-level result type: {type(value).__name__}. "
            "Allowed top-level values are String, Number, Boolean, Object, Array, or Null."
        )


def _normalize_expected_type(expected_type: str) -> str:
    etype = expected_type.strip()
    low = etype.lower()
    simple_types = {
        "string": "String",
        "number": "Number",
        "boolean": "Boolean",
        "object": "Object",
        "null": "Null",
        "any": "Any",
    }
    if low in simple_types:
        return simple_types[low]
    if low.startswith("array<") and low.endswith(">"):
        inner = etype[etype.find("<") + 1 : -1].strip()
        if not inner:
            raise ContractError(f"Unsupported expected type: {expected_type}")
        return f"Array<{_normalize_expected_type(inner)}>"
    return etype


def _validate_expected_type(expected_type: str, value: Any, path: str = "") -> None:
    etype = _normalize_expected_type(expected_type)
    if not etype or etype.lower() == "any":
        return

    value = normalize_output_value(value)

    if etype.startswith("Array<") and etype.endswith(">"):
        inner_type = etype[6:-1].strip()
        if not isinstance(value, list):
            raise ContractError(
                f"CodeExec contract mismatch at {path or 'value'}: expected type {etype}, got {infer_actual_type(value)}"
            )
        for index, item in enumerate(value):
            child_path = f"{path}[{index}]" if path else f"[{index}]"
            _validate_expected_type(inner_type, item, child_path)
        return

    actual_type = infer_actual_type(value)
    if etype == "String":
        valid = isinstance(value, str)
    elif etype == "Number":
        valid = _is_number(value)
    elif etype == "Boolean":
        valid = isinstance(value, bool)
    elif etype == "Object":
        valid = isinstance(value, dict)
    elif etype == "Null":
        valid = value is None
    else:
        raise ContractError(f"Unsupported expected type: {expected_type}")

    if not valid:
        raise ContractError(
            f"CodeExec contract mismatch at {path or 'value'}: expected type {etype}, got {actual_type}"
        )


def build_code_exec_contract(outputs: Mapping[str, Any], raw_result: Any) -> dict[str, Any]:
    business_name, business_meta = select_business_output(outputs)
    expected_type = ""
    if isinstance(business_meta, Mapping):
        expected_type = str(business_meta.get("type") or "")

    normalized_value = normalize_output_value(raw_result)
    _validate_top_level_value_domain(normalized_value)
    _validate_expected_type(expected_type, normalized_value)

    return {
        "business_output": business_name,
        "value": normalized_value,
        "actual_type": infer_actual_type(normalized_value),
        "content": render_canonical_content(normalized_value),
    }

File: internal/agent/tool/code_exec_contract.go
//
//  Copyright 2026 The InfiniFlow Authors. All Rights Reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

package tool

import (
	"encoding/json"
	"fmt"
	"math"
	"strings"
)

var codeExecSystemOutputKeys = map[string]struct{}{
	"content":             {},
	"actual_type":         {},
	"attachments":         {},
	"_ERROR":              {},
	"_ARTIFACTS":          {},
	"_ATTACHMENT_CONTENT": {},
	"raw_result":          {},
	"_created_time":       {},
	"_elapsed_time":       {},
}

type CodeExecContract struct {
	BusinessOutput string
	Value          any
	ActualType     string
	Content        string
}

func BuildCodeExecContract(outputs map[string]any, rawResult any) (*CodeExecContract, error) {
	businessName, businessMeta, err := selectCodeExecBusinessOutput(outputs)
	if err != nil {
		return nil, err
	}

	normalizedValue := NormalizeCodeExecOutputValue(rawResult)
	if err := validateCodeExecTopLevelValueDomain(normalizedValue); err != nil {
		return nil, err
	}

	expectedType := ""
	if meta, ok := businessMeta.(map[string]any); ok {
		expectedType = strings.TrimSpace(fmt.Sprint(meta["type"]))
	}
	if err := validateCodeExecExpectedType(expectedType, normalizedValue, ""); err != nil {
		return nil, err
	}

	return &CodeExecContract{
		BusinessOutput: businessName,
		Value:          normalizedValue,
		ActualType:     InferCodeExecActualType(normalizedValue),
		Content:        RenderCodeExecCanonicalContent(normalizedValue),
	}, nil
}

func NormalizeCodeExecOutputValue(value any) any {
	switch v := value.(type) {
	case []any:
		out := make([]any, 0, len(v))
		for _, item := range v {
			out = append(out, NormalizeCodeExecOutputValue(item))
		}
		return out
	case map[string]any:
		out := make(map[string]any, len(v))
		for key, item := range v {
			out[key] = NormalizeCodeExecOutputValue(item)
		}
		return out
	default:
		return v
	}
}

func InferCodeExecActualType(value any) string {
	value = NormalizeCodeExecOutputValue(value)
	switch v := value.(type) {
	case nil:
		return "Null"
	case bool:
		return "Boolean"
	case string:
		return "String"
	case map[string]any:
		return "Object"
	case []any:
		if len(v) == 0 {
			return "Array<Any>"
		}
		first := InferCodeExecActualType(v[0])
		for _, item := range v[1:] {
			if InferCodeExecActualType(item) != first {
				return "Array<Any>"
			}
		}
		return "Array<" + first + ">"
	case json.Number:
		return "Number"
	case float64:
		if math.IsNaN(v) || math.IsInf(v, 0) {
			return "Any"
		}
		return "Number"
	case float32:
		if math.IsNaN(float64(v)) || math.IsInf(float64(v), 0) {
			return "Any"
		}
		return "Number"
	case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64:
		return "Number"
	default:
		return "Any"
	}
}

func RenderCodeExecCanonicalContent(value any) string {
	value = NormalizeCodeExecOutputValue(value)
	switch v := value.(type) {
	case nil:
		return ""
	case string:
		return v
	case map[string]any, []any:
		b, err := json.MarshalIndent(v, "", "  ")
		if err != nil {
			return fmt.Sprint(v)
		}
		return string(b)
	default:
		return fmt.Sprint(v)
	}
}

func selectCodeExecBusinessOutput(outputs map[string]any) (string, any, error) {
	if len(outputs) == 1 {
		for name, meta := range outputs {
			if err := validateCodeExecBusinessOutputName(name); err != nil {
				return "", nil, err
			}
			return name, meta, nil
		}
	}

	var (
		businessName string
		businessMeta any
		count        int
	)
	for name, meta := range outputs {
		if _, reserved := codeExecSystemOutputKeys[name]; reserved {
			continue
		}
		count++
		businessName = name
		businessMeta = meta
	}
	if count != 1 {
		return "", nil, fmt.Errorf("CodeExec contract must contain exactly one business output, got %d", count)
	}
	if err := validateCodeExecBusinessOutputName(businessName); err != nil {
		return "", nil, err
	}
	return businessName, businessMeta, nil
}

func validateCodeExecBusinessOutputName(name string) error {
	name = strings.TrimSpace(name)
	if name == "" {
		return fmt.Errorf("CodeExec business output name must not be empty")
	}
	if _, reserved := codeExecSystemOutputKeys[name]; reserved {
		return fmt.Errorf("CodeExec reserved output name is not allowed: %s", name)
	}
	if strings.Contains(name, ".") {
		return fmt.Errorf("CodeExec business output name must not contain '.': %s", name)
	}
	return nil
}

func validateCodeExecTopLevelValueDomain(value any) error {
	switch value.(type) {
	case nil, bool, string, map[string]any, []any, json.Number,
		float64, float32, int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64:
		return nil
	default:
		return fmt.Errorf(
			"CodeExec unsupported top-level result type: %T. Allowed top-level values are String, Number, Boolean, Object, Array, or Null.",
			value,
		)
	}
}

func validateCodeExecExpectedType(expectedType string, value any, path string) error {
	etype, err := normalizeCodeExecExpectedType(expectedType)
	if err != nil {
		return err
	}
	if etype == "" || strings.EqualFold(etype, "Any") {
		return nil
	}

	value = NormalizeCodeExecOutputValue(value)
	if strings.HasPrefix(etype, "Array<") && strings.HasSuffix(etype, ">") {
		list, ok := value.([]any)
		if !ok {
			return fmt.Errorf(
				"CodeExec contract mismatch at %s: expected type %s, got %s",
				codeExecPathOrValue(path), etype, InferCodeExecActualType(value),
			)
		}
		innerType := strings.TrimSpace(etype[len("Array<") : len(etype)-1])
		for i, item := range list {
			childPath := fmt.Sprintf("[%d]", i)
			if path != "" {
				childPath = path + childPath
			}
			if err := validateCodeExecExpectedType(innerType, item, childPath); err != nil {
				return err
			}
		}
		return nil
	}

	valid := false
	switch etype {
	case "String":
		_, valid = value.(string)
	case "Number":
		switch value.(type) {
		case json.Number, float64, float32, int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64:
			valid = true
		}
	case "Boolean":
		_, valid = value.(bool)
	case "Object":
		_, valid = value.(map[string]any)
	case "Null":
		valid = value == nil
	default:
		return fmt.Errorf("Unsupported expected type: %s", expectedType)
	}
	if valid {
		return nil
	}
	return fmt.Errorf(
		"CodeExec contract mismatch at %s: expected type %s, got %s",
		codeExecPathOrValue(path), etype, InferCodeExecActualType(value),
	)
}

func normalizeCodeExecExpectedType(expectedType string) (string, error) {
	etype := strings.TrimSpace(expectedType)
	if etype == "" {
		return "", nil
	}
	switch strings.ToLower(etype) {
	case "string":
		return "String", nil
	case "number":
		return "Number", nil
	case "boolean":
		return "Boolean", nil
	case "object":
		return "Object", nil
	case "null":
		return "Null", nil
	case "any":
		return "Any", nil
	}
	low := strings.ToLower(etype)
	if strings.HasPrefix(low, "array<") && strings.HasSuffix(etype, ">") {
		inner := strings.TrimSpace(etype[len("Array<") : len(etype)-1])
		if inner == "" {
			return "", fmt.Errorf("Unsupported expected type: %s", expectedType)
		}
		normalizedInner, err := normalizeCodeExecExpectedType(inner)
		if err != nil {
			return "", err
		}
		return "Array<" + normalizedInner + ">", nil
	}
	return etype, nil
}

func codeExecPathOrValue(path string) string {
	if strings.TrimSpace(path) == "" {
		return "value"
	}
	return path
}
